src/nnet3bin/nnet3-discriminative-merge-egs.cc

```diff
@@ -0,0 +1,102 @@
+// nnet3bin/nnet3-discriminative-merge-egs.cc
+
+// Copyright 2012-2015  Johns Hopkins University (author:  Daniel Povey)
+//           2014-2015  Vimal Manohar
+
+// See ../../COPYING for clarification regarding multiple authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#include "base/kaldi-common.h"
+#include "util/common-utils.h"
+#include "hmm/transition-model.h"
+#include "nnet3/nnet-discriminative-example.h"
+
+
+int main(int argc, char *argv[]) {
+  try {
+    using namespace kaldi;
+    using namespace kaldi::nnet3;
+    typedef kaldi::int32 int32;
+    typedef kaldi::int64 int64;
+
+    const char *usage =
+        "This copies nnet3 discriminative training examples from input to output, merging them\n"
+        "into composite examples.  The --minibatch-size option controls how many egs\n"
+        "are merged into a single output eg.\n"
+        "\n"
+        "Usage:  nnet3-discriminative-egs [options] <egs-rspecifier> <egs-wspecifier>\n"
+        "e.g.\n"
+        "nnet3-discriminative-merge-egs --minibatch-size=128 ark:1.degs ark:- | nnet3-discriminative-train ... \n"
+        "See also nnet3-discriminative-copy-egs\n";
+
+    bool compress = false;
+    int32 minibatch_size = 64;
+
+    ParseOptions po(usage);
+    po.Register("minibatch-size", &minibatch_size, "Target size of minibatches "
+                "when merging (see also --measure-output-frames)");
+    po.Register("compress", &compress, "If true, compress the output examples "
+                "(not recommended unless you are writing to disk");
+
+    po.Read(argc, argv);
+
+    if (po.NumArgs() != 2) {
+      po.PrintUsage();
+      exit(1);
+    }
+
+    std::string examples_rspecifier = po.GetArg(1),
+        examples_wspecifier = po.GetArg(2);
+
+    SequentialNnetDiscriminativeExampleReader example_reader(examples_rspecifier);
+    NnetDiscriminativeExampleWriter example_writer(examples_wspecifier);
+
+    std::vector<NnetDiscriminativeExample> examples;
+    examples.reserve(minibatch_size);
+
+    int64 num_read = 0, num_written = 0;
+    while (!example_reader.Done()) {
+      const NnetDiscriminativeExample &cur_eg = example_reader.Value();
+      examples.resize(examples.size() + 1);
+      examples.back() = cur_eg;
+
+      bool minibatch_ready =
+          static_cast<int32>(examples.size()) >= minibatch_size;
```

Comment:
Generally we like the number of sequences in the minibatch to be a power of
two, and preferably a multiple of 64 (for reasons relating to NVidia board
architecture).  This is easier to ensure if we set it  absolutely, not as a
number of frames.
Dan

On Wed, Jan 6, 2016 at 12:07 AM, Vimal Manohar <notifications@github.com>
wrote:

> In src/nnet3bin/nnet3-discriminative-merge-egs.cc
> <https://github.com/danpovey/kaldi/pull/3#discussion_r48934967>:
>
> > +        examples_wspecifier = po.GetArg(2);
> > +
> > +    SequentialNnetDiscriminativeExampleReader example_reader(examples_rspecifier);
> > +    NnetDiscriminativeExampleWriter example_writer(examples_wspecifier);
> > +
> > +    std::vector<NnetDiscriminativeExample> examples;
> > +    examples.reserve(minibatch_size);
> > +
> > +    int64 num_read = 0, num_written = 0;
> > +    while (!example_reader.Done()) {
> > +      const NnetDiscriminativeExample &cur_eg = example_reader.Value();
> > +      examples.resize(examples.size() + 1);
> > +      examples.back() = cur_eg;
> > +
> > +      bool minibatch_ready =
> > +          static_cast<int32>(examples.size()) >= minibatch_size;
>
> minibatch_size is measured in terms of number of examples rather than the
> number of output frames. This is the same as in the chain code. Is there a
> reason why this is preferred? What should be the minibatch_size if the
> examples are 1.5s long? The default in chain code was 64.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/danpovey/kaldi/pull/3/files#r48934967>.
>